cylc/rose/jinja2_parser.py

```diff
@@ -0,0 +1,127 @@
+        Pair
+    )
+
+    _STRING_REGEX = re.compile(
+        '^[\'"].*[\'"]$'
+    )
+
+    def literal_eval(self, value):
+        r"""A jinja2 equivalent to Python's ast.literal_eval.
+
+        Parses and returns valid literals.
+
+        Raises exceptions for expressions.
+
+        Examples:
+            >>> parser = Parser()
+
+            # valid pythonic literals
+            >>> parser.literal_eval('"42"')
+            '42'
+            >>> parser.literal_eval('42')
+            42
+            >>> parser.literal_eval('(1,2,3)')
+            (1, 2, 3)
+            >>> parser.literal_eval('[1,2,3]')
+            [1, 2, 3]
+            >>> parser.literal_eval('{"a": 1, "b": 2, "c": 3}')
+            {'a': 1, 'b': 2, 'c': 3}
+            >>> parser.literal_eval('True')
+            True
+            >>> parser.literal_eval('None')
+
+            # valid jinja2 variants
+            >>> parser.literal_eval('042')
+            42
+            >>> parser.literal_eval('true')
+            True
+            >>> parser.literal_eval('1,2,3')
+            (1, 2, 3)
+            >>> parser.literal_eval('01,true,')
+            (1, True)
+
+            # multiline literals
+            >>> parser.literal_eval('"a"\n" string"')
+            'a string'
+            >>> parser.literal_eval('1,\n2,\n3')
+            (1, 2, 3)
+
+            # invalid examples
+            >>> parser.literal_eval('1 + 1')
+            Traceback (most recent call last):
+            Exception: Invalid literal: 1 + 1
+            <class 'jinja2.nodes.Add'>
+            >>> parser.literal_eval('range(5)')
+            Traceback (most recent call last):
+            Exception: Invalid literal: range(5)
+            <class 'jinja2.nodes.Call'>
+            >>> parser.literal_eval('1 if True else 0')
+            Traceback (most recent call last):
+            Exception: Invalid literal: 1 if True else 0
+            <class 'jinja2.nodes.CondExpr'>
+
+            # quirks
+            >>> parser.literal_eval('{1,2,3}')  # Jinja2 cannot handle sets
+            Traceback (most recent call last):
+            jinja2.exceptions.TemplateSyntaxError: expected token ':', got ','
+
+        """
+        # pass string values through ast.literal_eval
+        # (jinja2 will peel back the quotes to get at what's inside)
+        value = value.strip()
+        if self._STRING_REGEX.match(value):
+            return python_literal_eval(value)
+        # dump this value into a Jinja2 template
+        templ = '{{ %s }}' % value
+        # check that this expression consists only of literals
+        ast = self.parse(templ)
+        stack = [ast]
+        while stack:
+            node = stack.pop()
+            if not isinstance(node, self._LITERAL_NODES):
+                raise Exception(
+                    f'Invalid literal: {value}'
+                    f'\n{type(node)}'
+                )
+            stack.extend([x for x in node.iter_child_nodes()])
```

Comment:
IMO this isn't very nice to read, but a few minutes looking at the jinja2 api docs and Stackoverflow have convinced me that there's no clear magic for doing this better. Jinja2 ast's look rather limited in their abilities.

Can I suggest that (depending on how balkanized you want the code) that you consider:
a) Moving this check into a function with a name such as `is_legal_jinja2(value)` or
b) The following slight refactor:
```suggestion

        # dump this value into a Jinja2 template
        # check that this expression consists only of literals
        templ = '{{ %s }}' % value
        ast = self.parse(templ)
        stack = [ast]
        while stack:
            node = stack.pop()
            if not isinstance(node, self._LITERAL_NODES):
                raise Exception(
                    f'Invalid literal: {value}'
                    f'\n{type(node)}'
                )
            stack.extend([x for x in node.iter_child_nodes()])

```

Reply:
I don't think there is a much clearer way to express this code, it's very short and already has a comment:code ratio of 53:15.

Reply:
> I don't think there is a much clearer way to express this code, it's very short and already has a comment:code ratio of 53:15.

You could replace the stack with `Node.find_all` and avoid reparsing the expression, but other than that there isn't much fat that can be cut out here:

```python
        # dump this value into a Jinja2 template
        templ = '{{ %s }}' % value
        # check that this expression consists only of literals
        ast = self.parse(templ)
        for node in ast.find_all(jinja2.nodes.Node):
            if not isinstance(node, self._LITERAL_NODES):
                raise Exception(
                    f'Invalid literal: {value}'
                    f'\n{type(node)}'
                )
        # evaluate it
        return self.from_string(ast).render()  # <- avoids reparsing the expression
```

(Note that there is also [`jinja2.Environment.compile_expression`](https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.Environment.compile_expression) API but it fails some of your more perverse tests, namely the tuple ones, unless protected with extra parentheses: `return self.compile_expression('(%s)' % value)()`).

I wonder, however, if the following jinja-less solution wouldn't provide you with close enough approximation:

```python
        value = value.strip()
        if value in ('true', 'false'):
            value = value.capitalize()
        value = '(%s)' % value
        return ast.literal_eval(value)
```

This only fails on octal numbers (how widespread are these?), but it does work for sets.